sources/osgShadow/ShadowMapAtlas.js

```diff
@@ -314,10 +382,60 @@ MACROUTILS.createPrototypeObject(
 
         // Defines the frustum from light param.
         //
-        cullShadowCasting: function(cullVisitor) {
+        cullShadowCasting: function(cullVisitor, bbox) {
+            this._cameraClear.accept(cullVisitor);
+
             for (var i = 0, l = this._shadowMaps.length; i < l; i++) {
-                this._shadowMaps[i].cullShadowCasting(cullVisitor);
+                this._shadowMaps[i].cullShadowCasting(cullVisitor, bbox);
+            }
+        },
+
+        removeShadowMap: function(shadowMap) {
+            if (this._shadowMaps.length > 0) {
+                var idx = this._shadowMaps.indexOf(shadowMap);
+                if (idx !== -1) {
+                    if (this._shadowMaps[idx].valid()) {
+                        this._shadowMaps[idx].cleanSceneGraph(true);
+                    }
+                    this._shadowMaps.splice(idx, 1);
+
+                    var lightNumberArray = this._texture.getLightNumberArray();
+                    idx = lightNumberArray.indexOf(shadowMap.getLightNumber());
+                    if (idx !== -1) lightNumberArray.splice(idx, 1);
+
+                    var light = shadowMap.getLight();
+                    idx = this._lights.indexOf(light);
+                    if (idx !== -1) this._lights.splice(idx, 1);
+
+                    this.recomputeViewports();
+                }
+            }
+        },
+
+        addShadowMap: function(shadowMap) {
+            if (this._shadowMaps.length > 0) {
+                if (this._shadowMaps.indexOf(shadowMap) !== -1) return;
+            }
+
+            this._shadowMaps.push(shadowMap);
+
+            var light = shadowMap.getLight();
+            if (this._lights.indexOf(light) === -1) {
+                this._lights.push(light);
+            } else {
+                console.assert(false);
             }
+
+            if (this._shadowedScene) shadowMap.setShadowedScene(this._shadowedScene);
+
+            var lightNumberArray = this._texture.getLightNumberArray();
+            if (lightNumberArray.indexOf(shadowMap.getLightNumber()) === -1) {
+                lightNumberArray.push(light.getLightNumber());
+            } else {
+                console.assert(false);
```

Comment:
double uncool

Reply:
ha double sorry 
that debug code to double check error in lightnumber/index was helpful, but definitely not needed now